Bazel crashes on ActionCache replies with directory trees

Since [PR-1870](https://github.com/buildfarm/buildfarm/pull/1870), our Bazel client (7.2.1 and 7.4.0rc4 and 7.4.0) crash with the following stacktrace:

```
(20:01:49) FATAL: bazel crashed due to an internal error. Printing stack trace:
java.lang.RuntimeException: Unrecoverable error while evaluating node 'ActionLookupData0{actionLookupKey=ConfiguredTargetKey{label=//<REDACTED>, config=BuildConfigurationKey[bfedefdc22aebf3cda9247404ac1b7b3bff251e71d31d0b7f5766285194d085b]}, actionIndex=0}' (requested by nodes 'File:[[<execution_root>]bazel-out/k8-fastbuild/bin]<REDACTED>/path/to/dir')
	at com.google.devtools.build.skyframe.AbstractParallelEvaluator$Evaluate.run(AbstractParallelEvaluator.java:550)
	at com.google.devtools.build.lib.concurrent.AbstractQueueVisitor$WrappedRunnable.run(AbstractQueueVisitor.java:414)
	at java.base/java.util.concurrent.ForkJoinTask$RunnableExecuteAction.exec(ForkJoinTask.java:1423)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:387)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1312)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1843)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1808)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:188)
Caused by: java.lang.NullPointerException
	at com.google.common.base.Preconditions.checkNotNull(Preconditions.java:903)
	at com.google.devtools.build.lib.remote.RemoteExecutionService.parseDirectory(RemoteExecutionService.java:1017)
	at com.google.devtools.build.lib.remote.RemoteExecutionService.parseActionResultMetadata(RemoteExecutionService.java:1058)
	at com.google.devtools.build.lib.remote.RemoteExecutionService.downloadOutputs(RemoteExecutionService.java:1137)
	at com.google.devtools.build.lib.remote.RemoteSpawnRunner.downloadAndFinalizeSpawnResult(RemoteSpawnRunner.java:454)
	at com.google.devtools.build.lib.remote.RemoteSpawnRunner.exec(RemoteSpawnRunner.java:224)
	at com.google.devtools.build.lib.exec.AbstractSpawnStrategy.exec(AbstractSpawnStrategy.java:158)
	at com.google.devtools.build.lib.exec.AbstractSpawnStrategy.exec(AbstractSpawnStrategy.java:118)
	at com.google.devtools.build.lib.exec.SpawnStrategyResolver.exec(SpawnStrategyResolver.java:45)
	at com.google.devtools.build.lib.analysis.actions.SpawnAction.execute(SpawnAction.java:261)
	at com.google.devtools.build.lib.skyframe.SkyframeActionExecutor$ActionRunner.executeAction(SkyframeActionExecutor.java:1159)
	at com.google.devtools.build.lib.skyframe.SkyframeActionExecutor$ActionRunner.run(SkyframeActionExecutor.java:1076)
	at com.google.devtools.build.lib.skyframe.ActionExecutionState.runStateMachine(ActionExecutionState.java:165)
	at com.google.devtools.build.lib.skyframe.ActionExecutionState.getResultOrDependOnFuture(ActionExecutionState.java:94)
	at com.google.devtools.build.lib.skyframe.SkyframeActionExecutor.executeAction(SkyframeActionExecutor.java:573)
	at com.google.devtools.build.lib.skyframe.ActionExecutionFunction.checkCacheAndExecuteIfNeeded(ActionExecutionFunction.java:859)
	at com.google.devtools.build.lib.skyframe.ActionExecutionFunction.computeInternal(ActionExecutionFunction.java:333)
	at com.google.devtools.build.lib.skyframe.ActionExecutionFunction.compute(ActionExecutionFunction.java:171)
	at com.google.devtools.build.skyframe.AbstractParallelEvaluator$Evaluate.run(AbstractParallelEvaluator.java:461)
```

--- a/src/main/java/build/buildfarm/common/DigestPath.java
+++ b/src/main/java/build/buildfarm/common/DigestPath.java
@@ -1,0 +1,20 @@
+// Copyright 2024 The Buildfarm Authors. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package build.buildfarm.common;
+
+import build.buildfarm.v1test.Digest;
+import java.nio.file.Path;
+
+public record DigestPath(Digest digest, Path path) {}

--- a/src/main/java/build/buildfarm/common/DigestUtil.java
+++ b/src/main/java/build/buildfarm/common/DigestUtil.java
@@ -59,12 +59,13 @@ public class DigestUtil {
 
   public static final Map<DigestFunction.Value, Digest> empty =
       ImmutableMap.of(
-          DigestFunction.Value.MD5, DigestUtil.forHash("MD5").compute(ByteString.empty()),
-          DigestFunction.Value.SHA1, DigestUtil.forHash("SHA1").compute(ByteString.empty()),
-          DigestFunction.Value.SHA256, DigestUtil.forHash("SHA256").compute(ByteString.empty()),
-          DigestFunction.Value.SHA384, DigestUtil.forHash("SHA384").compute(ByteString.empty()),
-          DigestFunction.Value.SHA512, DigestUtil.forHash("SHA512").compute(ByteString.empty()),
-          DigestFunction.Value.BLAKE3, DigestUtil.forHash("BLAKE3").compute(ByteString.empty()));
+          DigestFunction.Value.MD5, DigestUtil.forHash("MD5").computeImpl(ByteString.empty()),
+          DigestFunction.Value.SHA1, DigestUtil.forHash("SHA1").computeImpl(ByteString.empty()),
+          DigestFunction.Value.SHA256, DigestUtil.forHash("SHA256").computeImpl(ByteString.empty()),
+          DigestFunction.Value.SHA384, DigestUtil.forHash("SHA384").computeImpl(ByteString.empty()),
+          DigestFunction.Value.SHA512, DigestUtil.forHash("SHA512").computeImpl(ByteString.empty()),
+          DigestFunction.Value.BLAKE3,
+              DigestUtil.forHash("BLAKE3").computeImpl(ByteString.empty()));
 
   /** Type of hash function to use for digesting blobs. */
   // The underlying HashFunctions are immutable and thread safe.
@@ -203,8 +204,12 @@ public class DigestUtil {
 
   public Digest compute(ByteString blob) {
     if (blob.size() == 0) {
-      return Digest.getDefaultInstance();
+      return empty();
     }
+    return computeImpl(blob);
+  }
+
+  private Digest computeImpl(ByteString blob) {
     return buildDigest(computeHash(blob).toString(), blob.size(), getDigestFunction());
   }
 

--- a/src/main/java/build/buildfarm/common/function/IOConsumer.java
+++ b/src/main/java/build/buildfarm/common/function/IOConsumer.java
@@ -1,0 +1,37 @@
+// Copyright 2024 The Buildfarm Authors. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package build.buildfarm.common.function;
+
+import java.io.IOException;
+
+/**
+ * Represents an operation that accepts a single input argument and returns no result, and may throw
+ * an IOException, implying that calls may perform I/O. Unlike most other functional interfaces,
+ * {@code IOConsumer} is expected to operate via side-effects.
+ *
+ * <p>This is a <a href="package-summary.html">functional interface</a> whose functional method is
+ * {@link #accept(Object)}.
+ *
+ * @param <T> the type of the input to the operation
+ */
+@FunctionalInterface
+public interface IOConsumer<T> {
+  /**
+   * Performs this I/O operation on the given argument.
+   *
+   * @param t the input argument
+   */
+  void accept(T t) throws IOException;
+}

--- a/src/main/java/build/buildfarm/worker/shard/ShardWorkerContext.java
+++ b/src/main/java/build/buildfarm/worker/shard/ShardWorkerContext.java
@@ -28,14 +28,11 @@ import build.bazel.remote.execution.v2.Command;
 import build.bazel.remote.execution.v2.Compressor;
 import build.bazel.remote.execution.v2.DigestFunction;
 import build.bazel.remote.execution.v2.Directory;
-import build.bazel.remote.execution.v2.DirectoryNode;
 import build.bazel.remote.execution.v2.ExecutionStage;
-import build.bazel.remote.execution.v2.FileNode;
 import build.bazel.remote.execution.v2.Platform;
-import build.bazel.remote.execution.v2.SymlinkNode;
-import build.bazel.remote.execution.v2.Tree;
 import build.buildfarm.backplane.Backplane;
 import build.buildfarm.common.CommandUtils;
+import build.buildfarm.common.DigestPath;
 import build.buildfarm.common.DigestUtil;
 import build.buildfarm.common.DigestUtil.ActionKey;
 import build.buildfarm.common.DigestUtil.HashFunction;
@@ -50,6 +47,7 @@ import build.buildfarm.common.SystemProcessors;
 import build.buildfarm.common.Write;
 import build.buildfarm.common.config.BuildfarmConfigs;
 import build.buildfarm.common.config.ExecutionPolicy;
+import build.buildfarm.common.function.IOConsumer;
 import build.buildfarm.common.grpc.Retrier;
 import build.buildfarm.common.grpc.Retrier.Backoff;
 import build.buildfarm.instance.Instance;
@@ -68,7 +66,6 @@ import build.buildfarm.worker.resources.Claim;
 import build.buildfarm.worker.resources.LocalResourceSet;
 import build.buildfarm.worker.resources.ResourceDecider;
 import build.buildfarm.worker.resources.ResourceLimits;
-import com.google.common.annotations.VisibleForTesting;
 import com.google.common.base.Preconditions;
 import com.google.common.base.Throwables;
 import com.google.common.collect.ImmutableList;
@@ -87,17 +84,12 @@ import io.prometheus.client.Counter;
 import java.io.File;
 import java.io.IOException;
 import java.io.InputStream;
-import java.nio.file.FileVisitResult;
 import java.nio.file.Files;
 import java.nio.file.NoSuchFileException;
 import java.nio.file.Path;
-import java.nio.file.SimpleFileVisitor;
-import java.nio.file.attribute.BasicFileAttributes;
 import java.util.ArrayList;
-import java.util.Comparator;
 import java.util.List;
 import java.util.Map;
-import java.util.Stack;
 import java.util.logging.Level;
 import javax.annotation.Nullable;
 import lombok.extern.java.Log;
@@ -573,36 +565,6 @@ class ShardWorkerContext implements WorkerContext {
     }
   }
 
-  @VisibleForTesting
-  static class OutputDirectoryContext {
-    private final List<FileNode> files = new ArrayList<>();
-    private final List<DirectoryNode> directories = new ArrayList<>();
-    private final List<SymlinkNode> symlinks = new ArrayList<>();
-
-    void addFile(FileNode fileNode) {
-      files.add(fileNode);
-    }
-
-    void addDirectory(DirectoryNode directoryNode) {
-      directories.add(directoryNode);
-    }
-
-    void addSymlink(SymlinkNode symlinkNode) {
-      symlinks.add(symlinkNode);
-    }
-
-    Directory toDirectory() {
-      files.sort(Comparator.comparing(FileNode::getName));
-      directories.sort(Comparator.comparing(DirectoryNode::getName));
-      symlinks.sort(Comparator.comparing(SymlinkNode::getName));
-      return Directory.newBuilder()
-          .addAllFiles(files)
-          .addAllDirectories(directories)
-          .addAllSymlinks(symlinks)
-          .build();
-    }
-  }
-
   private void uploadOutputDirectory(
       ActionResult.Builder resultBuilder,
       DigestUtil digestUtil,
@@ -628,108 +590,27 @@ class ShardWorkerContext implements WorkerContext {
       return;
     }
 
-    Tree.Builder treeBuilder = Tree.newBuilder();
-    OutputDirectoryContext outputRoot = new OutputDirectoryContext();
-    Files.walkFileTree(
-        outputDirPath,
-        new SimpleFileVisitor<>() {
-          OutputDirectoryContext currentDirectory = null;
-          final Stack<OutputDirectoryContext> path = new Stack<>();
-
-          @Override
-          public FileVisitResult visitFile(Path file, BasicFileAttributes attrs)
-              throws IOException {
-            if (configs.getWorker().isCreateSymlinkOutputs() && attrs.isSymbolicLink()) {
-              visitSymbolicLink(file);
-            } else {
-              visitRegularFile(file);
-            }
-            return FileVisitResult.CONTINUE;
+    IOConsumer<DigestPath> fileObserver =
+        digestPath -> {
+          Digest digest = digestPath.digest();
+          try {
+            insertFile(digest, digestPath.path());
+          } catch (InterruptedException e) {
+            throw new IOException(e);
+          } catch (EntryLimitException e) {
+            preconditionFailure
+                .addViolationsBuilder()
+                .setType(entrySizeViolationType)
+                .setSubject("blobs/" + DigestUtil.toString(digest))
+                .setDescription(
+                    "An output could not be uploaded because it exceeded "
+                        + "the maximum size of an entry");
           }
-
-          private void visitSymbolicLink(Path file) throws IOException {
-            // TODO convert symlinks with absolute targets within execution root to relative ones
-            currentDirectory.addSymlink(
-                SymlinkNode.newBuilder()
-                    .setName(file.getFileName().toString())
-                    .setTarget(Files.readSymbolicLink(file).toString())
-                    .build());
-          }
-
-          private void visitRegularFile(Path file) throws IOException {
-            Digest digest;
-            try {
-              // should we create symlink nodes in output?
-              // is buildstream trying to execute in a specific container??
-              // can get to NSFE for nonexistent symlinks
-              // can fail outright for a symlink to a directory
-              digest = digestUtil.compute(file);
-            } catch (NoSuchFileException e) {
-              log.log(
-                  Level.SEVERE,
-                  format(
-                      "error visiting file %s under output dir %s",
-                      outputDirPath.relativize(file), outputDirPath.toAbsolutePath()),
-                  e);
-              return;
-            }
-
-            // should we cast to PosixFilePermissions and do gymnastics there for executable?
-
-            // TODO symlink per revision proposal
-            currentDirectory.addFile(
-                FileNode.newBuilder()
-                    .setName(file.getFileName().toString())
-                    .setDigest(DigestUtil.toDigest(digest))
-                    .setIsExecutable(Files.isExecutable(file))
-                    .build());
-            try {
-              insertFile(digest, file);
-            } catch (InterruptedException e) {
-              throw new IOException(e);
-            } catch (EntryLimitException e) {
-              preconditionFailure
-                  .addViolationsBuilder()
-                  .setType(entrySizeViolationType)
-                  .setSubject("blobs/" + DigestUtil.toString(digest))
-                  .setDescription(
-                      "An output could not be uploaded because it exceeded the maximum size of an"
-                          + " entry");
-            }
-          }
-
-          @Override
-          public FileVisitResult preVisitDirectory(Path dir, BasicFileAttributes attrs) {
-            path.push(currentDirectory);
-            if (dir.equals(outputDirPath)) {
-              currentDirectory = outputRoot;
-            } else {
-              currentDirectory = new OutputDirectoryContext();
-            }
-            return FileVisitResult.CONTINUE;
-          }
-
-          @Override
-          public FileVisitResult postVisitDirectory(Path dir, IOException exc) {
-            OutputDirectoryContext parentDirectory = path.pop();
-            Directory directory = currentDirectory.toDirectory();
-            if (parentDirectory == null) {
-              treeBuilder.setRoot(directory);
-            } else {
-              parentDirectory.addDirectory(
-                  DirectoryNode.newBuilder()
-                      .setName(dir.getFileName().toString())
-                      // FIXME make one digestUtil for all
-                      .setDigest(DigestUtil.toDigest(digestUtil.compute(directory)))
-                      .build());
-              treeBuilder.addChildren(directory);
-            }
-            currentDirectory = parentDirectory;
-            return FileVisitResult.CONTINUE;
-          }
-        });
-    Tree tree = treeBuilder.build();
-    ByteString treeBlob = tree.toByteString();
+        };
+    TreeWalker treeWalker =
+        new TreeWalker(configs.getWorker().isCreateSymlinkOutputs(), digestUtil, fileObserver);
+    Files.walkFileTree(outputDirPath, treeWalker);
+    ByteString treeBlob = treeWalker.getTree().toByteString();
     Digest treeDigest = digestUtil.compute(treeBlob);
     insertBlob(treeDigest, treeBlob);
     resultBuilder

--- a/src/main/java/build/buildfarm/worker/shard/TreeWalker.java
+++ b/src/main/java/build/buildfarm/worker/shard/TreeWalker.java
@@ -1,0 +1,178 @@
+// Copyright 2024 The Buildfarm Authors. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package build.buildfarm.worker.shard;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+import static java.lang.String.format;
+
+import build.bazel.remote.execution.v2.Directory;
+import build.bazel.remote.execution.v2.DirectoryNode;
+import build.bazel.remote.execution.v2.FileNode;
+import build.bazel.remote.execution.v2.SymlinkNode;
+import build.bazel.remote.execution.v2.Tree;
+import build.buildfarm.common.DigestPath;
+import build.buildfarm.common.DigestUtil;
+import build.buildfarm.common.function.IOConsumer;
+import build.buildfarm.v1test.Digest;
+import java.io.IOException;
+import java.nio.file.FileVisitResult;
+import java.nio.file.Files;
+import java.nio.file.NoSuchFileException;
+import java.nio.file.Path;
+import java.nio.file.SimpleFileVisitor;
+import java.nio.file.attribute.BasicFileAttributes;
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Stack;
+import java.util.logging.Level;
+import lombok.extern.java.Log;
+
+/** May be used multiple times, but not threadsafe */
+@Log
+class TreeWalker extends SimpleFileVisitor<Path> {
+  private static final class OutputDirectoryContext {
+    private final List<FileNode> files = new ArrayList<>();
+    private final List<DirectoryNode> directories = new ArrayList<>();
+    private final List<SymlinkNode> symlinks = new ArrayList<>();
+
+    void addFile(FileNode fileNode) {
+      files.add(fileNode);
+    }
+
+    void addDirectory(DirectoryNode directoryNode) {
+      directories.add(directoryNode);
+    }
+
+    void addSymlink(SymlinkNode symlinkNode) {
+      symlinks.add(symlinkNode);
+    }
+
+    Directory toDirectory() {
+      files.sort(Comparator.comparing(FileNode::getName));
+      directories.sort(Comparator.comparing(DirectoryNode::getName));
+      symlinks.sort(Comparator.comparing(SymlinkNode::getName));
+      return Directory.newBuilder()
+          .addAllFiles(files)
+          .addAllDirectories(directories)
+          .addAllSymlinks(symlinks)
+          .build();
+    }
+  }
+
+  private final Stack<OutputDirectoryContext> path = new Stack<>();
+  private final boolean createSymlinkOutputs;
+  private final DigestUtil digestUtil;
+  private final IOConsumer<DigestPath> fileObserver;
+  private Tree.Builder treeBuilder = null;
+  private OutputDirectoryContext currentDirectory = null;
+  private Tree tree = null;
+  private Path root = null;
+
+  TreeWalker(
+      boolean createSymlinkOutputs, DigestUtil digestUtil, IOConsumer<DigestPath> fileObserver) {
+    this.createSymlinkOutputs = createSymlinkOutputs;
+    this.digestUtil = digestUtil;
+    this.fileObserver = fileObserver;
+  }
+
+  Tree getTree() {
+    // only valid after peforming a walk
+    return checkNotNull(tree);
+  }
+
+  @Override
+  public FileVisitResult visitFile(Path file, BasicFileAttributes attrs) throws IOException {
+    if (createSymlinkOutputs && attrs.isSymbolicLink()) {
+      visitSymbolicLink(file);
+    } else {
+      visitRegularFile(file);
+    }
+    return FileVisitResult.CONTINUE;
+  }
+
+  private void visitSymbolicLink(Path file) throws IOException {
+    // TODO convert symlinks with absolute targets within execution root to relative ones
+    currentDirectory.addSymlink(
+        SymlinkNode.newBuilder()
+            .setName(file.getFileName().toString())
+            .setTarget(Files.readSymbolicLink(file).toString())
+            .build());
+  }
+
+  private void visitRegularFile(Path file) throws IOException {
+    Digest digest;
+    try {
+      // should we create symlink nodes in output?
+      // is buildstream trying to execute in a specific container??
+      // can get to NSFE for nonexistent symlinks
+      // can fail outright for a symlink to a directory
+      digest = digestUtil.compute(file);
+    } catch (NoSuchFileException e) {
+      log.log(
+          Level.SEVERE,
+          format(
+              "error visiting file %s under output dir %s",
+              root.relativize(file), root.toAbsolutePath()),
+          e);
+      return;
+    }
+
+    // should we cast to PosixFilePermissions and do gymnastics there for executable?
+
+    // TODO symlink per revision proposal
+    currentDirectory.addFile(
+        FileNode.newBuilder()
+            .setName(file.getFileName().toString())
+            .setDigest(DigestUtil.toDigest(digest))
+            .setIsExecutable(Files.isExecutable(file))
+            .build());
+    fileObserver.accept(new DigestPath(digest, file));
+  }
+
+  @Override
+  public FileVisitResult preVisitDirectory(Path dir, BasicFileAttributes attrs) {
+    path.push(currentDirectory);
+    if (currentDirectory == null) {
+      // reset state when at root
+      treeBuilder = Tree.newBuilder();
+      root = dir;
+    }
+    currentDirectory = new OutputDirectoryContext();
+    return FileVisitResult.CONTINUE;
+  }
+
+  @Override
+  public FileVisitResult postVisitDirectory(Path dir, IOException exc) {
+    OutputDirectoryContext parentDirectory = path.pop();
+    Directory directory = currentDirectory.toDirectory();
+    if (parentDirectory == null) {
+      treeBuilder.setRoot(directory);
+      tree = treeBuilder.build();
+      treeBuilder = null;
+      root = null;
+    } else {
+      parentDirectory.addDirectory(
+          DirectoryNode.newBuilder()
+              .setName(dir.getFileName().toString())
+              // FIXME make one digestUtil for all
+              .setDigest(DigestUtil.toDigest(digestUtil.compute(directory)))
+              .build());
+      treeBuilder.addChildren(directory);
+    }
+    currentDirectory = parentDirectory;
+    return FileVisitResult.CONTINUE;
+  }
+}

--- a/src/test/java/build/buildfarm/common/DigestUtilTest.java
+++ b/src/test/java/build/buildfarm/common/DigestUtilTest.java
@@ -105,10 +105,11 @@ public class DigestUtilTest {
   }
 
   @Test
-  public void computeEmptyIsDefault() {
+  public void computeEmptyIsCachedEmpty() {
     DigestUtil digestUtil = new DigestUtil(HashFunction.BLAKE3);
     Digest digest = digestUtil.compute(ByteString.empty());
-    assertThat(digest == Digest.getDefaultInstance()).isTrue();
+    // reference comparison
+    assertThat(digest == digestUtil.empty()).isTrue();
   }
 
   @Test(expected = IllegalArgumentException.class)

--- a/src/test/java/build/buildfarm/worker/shard/TreeWalkerTest.java
+++ b/src/test/java/build/buildfarm/worker/shard/TreeWalkerTest.java
@@ -1,0 +1,132 @@
+// Copyright 2024 The Buildfarm Authors. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package build.buildfarm.worker.shard;
+
+import static com.google.common.truth.Truth.assertThat;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+
+import build.bazel.remote.execution.v2.Directory;
+import build.bazel.remote.execution.v2.DirectoryNode;
+import build.bazel.remote.execution.v2.FileNode;
+import build.bazel.remote.execution.v2.SymlinkNode;
+import build.bazel.remote.execution.v2.Tree;
+import build.buildfarm.common.DigestPath;
+import build.buildfarm.common.DigestUtil;
+import build.buildfarm.common.function.IOConsumer;
+import build.buildfarm.v1test.Digest;
+import com.google.common.collect.Iterables;
+import com.google.common.jimfs.Jimfs;
+import com.google.protobuf.ByteString;
+import java.io.IOException;
+import java.io.OutputStream;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.List;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+
+@RunWith(JUnit4.class)
+public class TreeWalkerTest {
+  private final Path root = Iterables.getOnlyElement(Jimfs.newFileSystem().getRootDirectories());
+
+  @Test
+  public void treeShouldHaveHashedEmptyDirsInChildren() throws IOException {
+    // ensure that we have a clean topdir
+    Path treeRoot = root.resolve("tree_root");
+    Files.createDirectory(treeRoot);
+    Files.createDirectories(treeRoot.resolve("empty_subdir"));
+    DigestUtil digestUtil = DigestUtil.forHash("BLAKE3");
+    TreeWalker treeWalker =
+        new TreeWalker(/* createSymlinkOutputs= */ false, digestUtil, digestPath -> {});
+    Files.walkFileTree(treeRoot, treeWalker);
+    Tree tree = treeWalker.getTree();
+    DirectoryNode directoryNode = Iterables.getOnlyElement(tree.getRoot().getDirectoriesList());
+    assertThat(directoryNode.getName()).isEqualTo("empty_subdir");
+    Digest digest = digestUtil.toDigest(directoryNode.getDigest());
+    assertThat(digestUtil.compute(Iterables.getOnlyElement(tree.getChildrenList())))
+        .isEqualTo(digest);
+    assertThat(digest).isEqualTo(digestUtil.empty());
+    assertThat(digest.getHash().isEmpty()).isFalse();
+  }
+
+  @Test
+  public void visitsSymlinksAsFilesWhenConfigured() throws IOException {
+    DigestUtil digestUtil = DigestUtil.forHash("BLAKE3");
+    ByteString content = ByteString.copyFromUtf8("Hello, World");
+    Digest digest = digestUtil.compute(content);
+
+    IOConsumer<DigestPath> fileObserver = mock(IOConsumer.class);
+    TreeWalker treeWalker =
+        new TreeWalker(/* createSymlinkOutputs= */ false, digestUtil, fileObserver);
+    Path treeRoot = root.resolve("tree_root");
+    Files.createDirectory(treeRoot);
+    Path filePath = treeRoot.resolve("file");
+    try (OutputStream out = Files.newOutputStream(filePath)) {
+      content.writeTo(out);
+    }
+    Path symlinkPath = treeRoot.resolve("symlink_to_file");
+    Files.createSymbolicLink(symlinkPath, treeRoot.relativize(filePath));
+    Files.walkFileTree(treeRoot, treeWalker);
+    Tree tree = treeWalker.getTree();
+    Directory rootDirectory = tree.getRoot();
+    List<FileNode> files = rootDirectory.getFilesList();
+    assertThat(files.size()).isEqualTo(2);
+    assertThat(files.get(0).getDigest()).isEqualTo(files.get(1).getDigest());
+    assertThat(files.get(0).getName()).isEqualTo("file");
+    assertThat(files.get(1).getName()).isEqualTo("symlink_to_file");
+    assertThat(rootDirectory.getSymlinksCount()).isEqualTo(0);
+    verify(fileObserver, times(1)).accept(new DigestPath(digest, filePath));
+    verify(fileObserver, times(1)).accept(new DigestPath(digest, symlinkPath));
+  }
+
+  @Test
+  public void visitsSymlinksWhenConfigured() throws IOException {
+    DigestUtil digestUtil = DigestUtil.forHash("BLAKE3");
+
+    TreeWalker treeWalker =
+        new TreeWalker(/* createSymlinkOutputs= */ true, digestUtil, digestPath -> {});
+    Path treeRoot = root.resolve("tree_root");
+    Files.createDirectory(treeRoot);
+    Path filePath = treeRoot.resolve("file");
+    Files.createSymbolicLink(treeRoot.resolve("symlink"), treeRoot.relativize(filePath));
+    Files.walkFileTree(treeRoot, treeWalker);
+    Tree tree = treeWalker.getTree();
+    Directory rootDirectory = tree.getRoot();
+    SymlinkNode symlink = Iterables.getOnlyElement(rootDirectory.getSymlinksList());
+    assertThat(symlink.getName()).isEqualTo("symlink");
+    assertThat(symlink.getTarget()).isEqualTo("file");
+    assertThat(rootDirectory.getFilesCount()).isEqualTo(0);
+  }
+
+  @Test
+  public void ignoresDeadSymlinksWhenConfigured() throws IOException {
+    DigestUtil digestUtil = DigestUtil.forHash("BLAKE3");
+
+    TreeWalker treeWalker =
+        new TreeWalker(/* createSymlinkOutputs= */ false, digestUtil, digestPath -> {});
+    Path treeRoot = root.resolve("tree_root");
+    Files.createDirectory(treeRoot);
+    Path filePath = treeRoot.resolve("file");
+    Files.createSymbolicLink(treeRoot.resolve("dead_symlink"), treeRoot.relativize(filePath));
+    Files.walkFileTree(treeRoot, treeWalker);
+    Tree tree = treeWalker.getTree();
+    Directory rootDirectory = tree.getRoot();
+    assertThat(rootDirectory.getSymlinksCount()).isEqualTo(0);
+    assertThat(rootDirectory.getFilesCount()).isEqualTo(0);
+  }
+}

